### PR TITLE
Syseleven/add prometheus support

### PR DESCRIPTION
**Pull request by Arabus**
Thought this might be useful ;-)

### CODE REVIEW

File: templates/service.yaml [hunk 1]
```diff
@@ -30,6 +30,12 @@ spec:
 {{- if (and (eq .Values.service.type "NodePort") (not (empty .Values.service.nodePort))) }}
       nodePort: {{ .Values.service.nodePort }}
 {{- end }}
+      {{- if .Values.serviceMonitor.enabled }}
+    - port: {{ (split ':' .Values.configData.http.debug.addr)._1 }}
```
Comment (canterberry) [inline templates/service.yaml:34]: If we're discarding the IP address portion of `configData.http.debug.addr`, let's just rename `addr` to `port` and have that expect the port alone. Then, this line could just be `- port: {{ .Values.configData.http.debug.port }}`.

File: templates/servicemonitor.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+{{ if .Values.serviceMonitor.enabled }}
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: {{ include "docker-registry.fullname" . }}
+  labels:
+    {{- include "docker-registry.labels" . | nindent 4 }}
```
Comment (canterberry) [inline templates/servicemonitor.yaml:7]: :+1: I like this -- much cleaner than the hard-coded/duplicated labels in each of the other templates. Can this be extracted into a separate pull request, though? If this PR introduces a `docker-registry.labels` helper, I'd like to simultaneously adopt that helper consistently across all templates instead of applying to just one.

File: templates/servicemonitor.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+spec:
+  endpoints:
+    - port: debug
+      path: /metrics
```
Comment (canterberry) [inline templates/servicemonitor.yaml:11]: Does it make sense for this path to be configurable? Or is the path here hard-coded based on what the container already provides? Are there ways to configure this chart that might make the path invalid?

File: values.yaml [hunk 4]
```diff
@@ -107,6 +107,12 @@ configData:
     addr: :5000
     headers:
       X-Content-Type-Options: [nosniff]
+    # Enable this if you want prometheus metrics
+    #debug:
+    #  addr: 0.0.0.0:5001
+    #  prometheus:
+    #    enabled: true
+    #    path: /metrics
```
Comment (canterberry) [inline values.yaml:115]: `configData.http.debug.prometheus` is commented-out here but is not referenced anywhere in the templates.

File: templates/servicemonitor.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
+{{ if .Values.serviceMonitor.enabled }}
```
Comment (canterberry) [inline templates/servicemonitor.yaml:1]: The service monitor requires a metrics endpoint to be configured and available. This should not only check whether the chart is configured to enable the service monitor, but also whether there is anything to monitor. Basically, this template could use a `.Values.configData.http.debug.prometheus.enabled` check as well.

**Comment by canterberry** [comment]
Thanks @Arabus! Initial review pass complete. Thanks for this.
As I got immersed more deeply in review, it all started to fall into place, and some comments I had toward the beginning of the review I was able to delete. What's left is distilled to a few spots of minor feedback and adjustments.

To reiterate my inline comment regarding the new helpers: I like where you're going with those changes, but I do think it muddies up this PR a bit and deserves its own PR to tackle separately and adopt consistently for other templates as well, not just the new servicemonitor.yaml template. For now, just follow the labels/conventions adopted in the other templates, dated as they may be. Then we can address modernization as a whole in a separate thread.

File: values.yaml [hunk 6]
```diff
@@ -145,3 +151,8 @@ extraVolumes: []
 #        - key: cloudfront.pem
 #          path: cloudfront.pem
 #          mode: 511
+
+serviceMonitor:
+  enabled: false
```
Comment (canterberry) [inline values.yaml:156]: :memo: Need to document this new configuration value in the README.

File: templates/servicemonitor.yaml (hunk 2 above)
Comment (Arabus) [reply templates/servicemonitor.yaml:7]: Sounds reasonable - this was mainly part of my PR because I already used it in a helm chart I had written for the same purpose as this ( which I wrote because I thought this one had been abandoned). I can't claim credit for that though, it is a standard feature of modern helm charts when building them via `helm create`. 

I will extract it for the time being and create a separate PR based on on this one.

File: templates/servicemonitor.yaml (hunk 3 above)
Comment (Arabus) [reply templates/servicemonitor.yaml]: Integrated this, though checking for values in optional subdicts in helm is pita. Also setting global values from inside an if feels messy.

File: templates/servicemonitor.yaml (hunk 5 above)
Comment (Arabus) [reply templates/servicemonitor.yaml]: This was rather tricky because the complete `http.` subdict is optional and helm tends to fail with nilPointer exceptions whe you reference values whose parents aren't there. I found a workaround though i am not really happy with it:
`{{- if not ( and ( hasKey .Values "configData" ) ( hasKey .Values.configData "http" ) ( hasKey .Values.configData.http "debug" ) ( hasKey .Values.configData.http.debug "prometheus" ) ( hasKey .Values.configData.http.debug.prometheus "enabled" ) ( .Values.configData.http.debug.prometheus.enabled ) ) -}}`

File: values.yaml (hunk 6 above)
Comment (Arabus) [reply values.yaml:164]: Check, also added some lines to the values.yaml

File: templates/servicemonitor.yaml (hunk 5 above)
Comment (Arabus) [reply templates/servicemonitor.yaml]: scratch that ... go and does not use lazy evaluation

Comment (Arabus) [reply templates/servicemonitor.yaml]: ok i blew a whole working day on this and it appears there is no way to make helm lookup a nested dict value where the parent dict does not exist... The only viable solution seems to be stacking if clauses... i will put them in tommorow after i have gone and yelled at the helm devs a little...

Comment (canterberry) [reply templates/servicemonitor.yaml]: I think `dig` might be what you need here. http://masterminds.github.io/sprig/dicts.html

Comment (Arabus) [reply templates/servicemonitor.yaml]: Holy cow... I scrolled through the libraries the whole day and missed that one completely... ok that should solve the above approach. I have gone and rewritten the PR in the meantime though, going with a more "user centric" one and allowing people to enable the debug port / metrics service endpoint without enabling the serviceMonitor at all. 

I am currently finishing it a little more and will push it then.

Comment (Arabus) [reply templates/servicemonitor.yaml]: ok updated the PR - this has become somewhat of a code splurge so if this feels like too much at once I am happy to break it down into smaller PRs.

File: README.md [hunk 7]
```diff
@@ -86,6 +86,13 @@ their default values.
 | `ingress.tls`               | Ingress TLS configuration (YAML)                                                           | `[]`            |
 | `extraVolumeMounts`         | Additional volumeMounts to the registry container                                          | `[]`            |
 | `extraVolumes`              | Additional volumes to the pod                                                              | `[]`            |
+| `metrics.enabled`           | Enable the debug metrics interface of the registry                                         | `false`         |
+| `metrics.servicePortName    | The name of the service port for the debug interface                                       | `debug`         |
+| `metrics.servicePort        | The port number of the service port for the debug interface                                | `5001`          |
+| `metrics.targetPortName     | The name of the container port of the debug interface                                      | `debug`         |
+| `metrics.serviceMonitor     | Various Prometheus serviceMonitor resource configs                                         | `nil`           |
```
Comment (canterberry) [inline README.md:93]: Missing closing backticks on these four lines.

File: templates/deployment.yaml [hunk 8]
```diff
@@ -54,46 +54,54 @@ spec:
           - /etc/docker/registry/config.yml
           ports:
             - containerPort: 5000
+            {{- if or .Values.metrics.enabled .Values.configData.http.debug.addr }}
```
Comment (canterberry) [inline templates/deployment.yaml:57]: `or` should be `and` here.

File: templates/deployment.yaml [hunk 9]
```diff
@@ -54,46 +54,54 @@ spec:
...
+            - name: {{ .Values.metrics.targetPortName | default "debug" | quote }}
+              containerPort: {{ (split ":" .Values.configData.http.debug.addr)._1 }}
+            {{- else }}
+              {{- if .Values.metrics.serviceMonitor.enabled }}
+                {{- fail "You enabled the serviceMonitor without enabling the registry debug endpoint. Please set metrics.enabled: true or specify configData.http.debug.addr" }}
```
Comment (canterberry) [inline templates/deployment.yaml:62]: "or" should be "and" here as well.

Basically, `metrics.enabled` only makes sense if the registry is configured to emit metrics, so it effectively depends on `configData.http.debug.addr` being set.

File: templates/servicemonitor.yaml [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    matchNames:
+      - {{ .Release.Namespace | quote }}
+  {{- if .Values.metrics.serviceMonitor.sampleLimit }}
+  sampleLimit: {{ .Values.metrics.serviceMonitor.sampleLimit }}
+  {{- end }}
+  {{- if .Values.metrics.serviceMonitor.targetLimit }}
+  targetLimit: {{ .Values.metrics.serviceMonitor.targetLimit }}
+  {{- end }}
+{{- end }}
+{{- end }}
```
Comment (canterberry) [inline templates/servicemonitor.yaml:97]: This template feels like overkill to me. If every single value is configurable, then it offers little benefit over a user just writing a separate manifest and deploying that. Instead, I'd propose just writing the bare minimum necessary for a functional service monitor in this template, and reduce its configurability to the bare essentials. For fine-grained configurability, the user can disable this chart from deploying a service monitor and the user can deploy their own alongside this chart rather than through it.

File: values.yaml [hunk 11]
```diff
@@ -109,6 +111,19 @@ configData:
...
+    ## Set values here to override the defaults
+    # debug:
+      ## The address and port the debug server should listen on
+      # addr: 0.0.0.0:5001
+      ## Prometheus metrics specific configuration
+      # prometheus:
+          ## Do not disable this if you use metrics.enabled: true
+          # enabled: true
+          ## Make sure to adjust metrics.serviceMonitor.path accordingly
+          # path: /metrics
```
Comment (canterberry) [inline values.yaml:126]: Just thinking aloud and paraphrasing to make sure I understand how this works:

 * To enable metrics, all a user has to do is set `metrics.enabled: true` in the values provided to this chart. If that is set, then a default configuration (as described above) will be injected into the registry's configuration to activate the debug interface and expose it on a default port.
 * If `metrics.enabled: true` is set, and `configData` already has a debug interface configuration, then this chart will derive the relevant port/path/etc from that registry config.

If those assumptions are correct, that seems like reasonable behavior to me.

File: values.yaml [hunk 12]
```diff
@@ -147,3 +162,46 @@ extraVolumes: []
...
+##
+metrics:
+  enabled: true
+  ## The name of the service resources port for the metrics interface
+  servicePortName: debug
+  ## The service port of the services clusterIP
+  servicePort: 5001
+
+  ## The container port name of the registry pods
+  targetPortName: debug
```
Comment (canterberry) [inline values.yaml:177]: I think it's probably okay to be opinionated about what these port names are, rather than making them configurable. Besides, I'm pretty sure the port names are just an implementation detail anyway with no impact outside of how this chart internally "glues" the service/deployment/pod together.

**Comment by canterberry** [comment]
Closing this PR to tidy up the list of ongoing work. Please feel free to reopen or continue the discussion.
